Blog post: may 2024 embroider update

BlueCutOfficial opened the pull request.

BlueCutOfficial commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,155 @@
+successful and she made impactful contributions to the project over the past
+three months.
+
+## Last words
+
+We are now at the point of wrapping up the technical challenges. We should soon
+have a Vite demo that is a good approximation of the new way to write Ember apps
+and matches the future blueprint. The next phase will be all about communication
+and community work: we want this new fast Embroider + Vite experience to be the
+default experience for all Ember users.
```
Do we want to tease Chris' talk at EmberConf next week?

KevinBongart replied:
YES

KevinBongart replied:
We can update the blog post later with a link to Chris' talk, too

BlueCutOfficial replied:
I added one last paragraph, what do you think?

BlueCutOfficial commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+  image: ""
+tagline: |
+  <p>
+    This post covers the progress over the last six months on [the Embroider project](https://github.com/embroider-build/embroider) that has been made through Mainmatter's [Embroider Initiative](https://mainmatter.com/embroider-initiative/).
```
Change links to HTML

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+---
+
+In the
+[November 2023 blog post](https://mainmatter.com/blog/2023/11/16/embroider-initiative-progress-update/),
```
```suggestion
[blog post from November 2023](https://mainmatter.com/blog/2023/11/16/embroider-initiative-progress-update/),
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+[November 2023 blog post](https://mainmatter.com/blog/2023/11/16/embroider-initiative-progress-update/),
+[Chris Manson](https://github.com/mansona) introduced to you Embroider, the new
+build pipeline that compiles Ember apps into spec-compliant, modern JavaScript.
+If this is the first time you have heard about Embroider, we encourage you to
```
```suggestion
If this is the first time you hear about Embroider, we encourage you to
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+read the introduction of the November blog post, which explains what Embroider
+is in more detail and presents the Embroider Initiative.
+
+The main goal that we laid out in the last blog was to get our Ember apps
```
```suggestion
The main goal that was set in the last blog was to get our Ember apps
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+
+## Improving how Ember works with Vite
+
+Now that our challenge is no longer to get Ember working with Vite we need to
```
```suggestion
Now that our challenge is no longer to get Ember working with Vite, we need to
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+
+Now that our challenge is no longer to get Ember working with Vite we need to
+figure out how to optimize the Vite build and maintain the compatibility of all
+classic features. To understand how to optimize, we need to dive into the
```
```suggestion
classic features. To find a solution, we need to dive into the
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+Now that our challenge is no longer to get Ember working with Vite we need to
+figure out how to optimize the Vite build and maintain the compatibility of all
+classic features. To understand how to optimize, we need to dive into the
+challenge that an Ember build presents to us
```
```suggestion
challenge that an Ember build presents.
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+is because Ember performs some “magic” around resolving imports, relying on
+[AMD modules](https://github.com/emberjs/rfcs/pull/938) and other specific build
+actions, which are incompatible with the Vite resolver. Let’s take an example:
+if you generated an Ember app using the default blueprint, and you will get a
```
```suggestion
if you generated an Ember app using the default blueprint, you will get a
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+if you generated an Ember app using the default blueprint, and you will get a
+file `app/index.html` which contains a reference to the CSS stylesheet
+`“assets/vendor.css”`. This file is supposed to contain the vendor styles that
+are automatically provided by the Ember addons you use, but that files doesn’t
```
```suggestion
are automatically provided by the Ember addons you use. However, those files don’t
```

BlueCutOfficial replied:
```suggestion
are automatically provided by the Ember addons you use. However, this file doesn’t
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+file `app/index.html` which contains a reference to the CSS stylesheet
+`“assets/vendor.css”`. This file is supposed to contain the vendor styles that
+are automatically provided by the Ember addons you use, but that files doesn’t
+exist in your repo. Your `app/index.html` points to a file that is literally
```
```suggestion
exist in your repo: your `app/index.html` points to a file that is literally
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+`“assets/vendor.css”`. This file is supposed to contain the vendor styles that
+are automatically provided by the Ember addons you use, but that files doesn’t
+exist in your repo. Your `app/index.html` points to a file that is literally
+nowhere in your app files. At build time Ember will go through all the active
```
```suggestion
nowhere in your app files. At build time, Ember will go through all the active
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+exist in your repo. Your `app/index.html` points to a file that is literally
+nowhere in your app files. At build time Ember will go through all the active
+Ember addons used in your app and will emit the `assets/vendor.css` file at some
+point in the build pipeline. But this is an Ember-specific behavior that Vite
```
```suggestion
point in the build pipeline. This is an Ember-specific behavior that Vite
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+To solve this issue, when the build starts, Embroider first produces an
+intermediate Ember application that we call the “rewritten app” and outputs it
+to `node_modules/.embroider/rewritten-app`. The rewritten app is slightly
+different from the initial Ember app, in a way it can be consumed by Vite. For
```
```suggestion
different from the initial Ember app in that it can be consumed by Vite. For
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+different from the initial Ember app, in a way it can be consumed by Vite. For
+instance, it has an `index.html` at the root and it does contain the
+`assets/vendor.css`. This approach allows us to build real-world applications
+with Vite, but it has performance (and other) downsides: Vite prebuilds the app
```
```suggestion
with Vite, but also carries performance (and other) downsides: Vite prebuilds the app
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+with Vite, but it has performance (and other) downsides: Vite prebuilds the app
+dependencies for optimization purposes, but the rewritten app _is_ considered a
+dependency because it lives in a folder in node_modules. Therefore, changes in
+the Ember app that need to be reflected in the rewritten app imply changes in a
```
```suggestion
the Ember app that need to be reflected in the rewritten app, imply changes in a
```

BlueCutOfficial replied:
I am not sure about this one 🤔 
If you delete the part that qualified the changes (), it would just be: 
"changes in the Ember app (...) imply changes in the rewritten app"

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+started a wide technical topic called “inversion of control”. The idea is the
+following: instead of having Embroider produce a rewritten app, and then passing
+over to Vite once the rewritten app is ready, Vite takes the lead, and when it’s
+unable to resolve Ember-specific things, it asks Embroider to return the
```
```suggestion
unable to resolve Ember-specific issues, it asks Embroider to return the
```

BlueCutOfficial replied:
```suggestion
unable to resolve Ember-specific requests, it asks Embroider to return the
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+file `assets/vendor.css`, and this file is referenced in the `index.html`. Vite
+has no trouble consuming that. But now, we want to get rid of the rewritten app
+step. To invert the control, we use a special identifier in the `index.html`
+that starts with `@embroider`. For instance: `@embroider/core/vendor.css`. Then
```
```suggestion
that starts with `@embroider`. For instance: `@embroider/core/vendor.css`. Then,
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+the rewritten app. Among these tasks, we have the virtualization of several
+files, handling the public assets provided by the app and addons, and also
+maintaining the compatibility of classic addons with Embroider by getting
+functionalities like content-for to work with Vite. A wide part of this work is
```
```suggestion
functionalities like content-for to work with Vite. A wide part of this work has
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+files, handling the public assets provided by the app and addons, and also
+maintaining the compatibility of classic addons with Embroider by getting
+functionalities like content-for to work with Vite. A wide part of this work is
+done already: all the app entrypoints are virtualized, the new authoring format
```
```suggestion
already been done: all the app entry points are virtualized, the new authoring format
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+all of Embroider's test suite uses Vite to get the app built.
+
+For the new tests, we started to reorganize things in a way that allows us to
+assert each functionality behaves as expected in build mode and dev mode: the
```
```suggestion
assert that each functionality behaves as expected in build mode and dev mode: the
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+production. Using this new audit system also allows us to more easly investigate
+the contents of the build, e.g. you don’t know how Rollup is going to
+organize the files, so you don’t know where to look for a specific piece of code
+you want to verify made it into the build.
```
I don't get this last sentence, what were you trying to say? "you don’t know how Rollup is going to
organize the files, so you don’t know where to look for a specific piece of code
you want to verify **made it into the build**."

KevinBongart replied:
While I proposed a fix for two typos above, I also don't get that sentence. Is the situation where Rollup organize the files randomly the current situation, and improving the audit system fixes that by organizing the content of the build in a way that can be investigated? If so, it should be phrased explicitly, for instance:

"Before this improvement, Rollup would organize the files seemingly at random, which made it difficult to investigate the content of the build. With this new audit system, files are exactly where you'd expect them, so you can do x and y."

BlueCutOfficial replied:
In that context "the build" is the output ready-for-production folder that our pipeline _builds_ out of the source code. So it means "you want to verify a specific piece of code that is present in your source code was indeed written somewhere in the build by the pipeline (but you can't know the name of the file it was written in)". Literally, imagine a piece of code moving from one place to made it to somewhere else ^^ 

Note that Chris phrased the sentence this way, it was part of his suggestion, and as a developer, I didn't have any trouble with it, but if you think it can be a problem maybe we can try to change it a little bit.

emmasofiah2o replied:
My issue is with this part of the sentence, where 4 different verbs are used. I can't tell what the subjects are and it sounds very confusing. Maybe you can try typing it in French? I am not sure how to fix it either because I don't really understand the context. 

"so you **don’t know** where to look for a specific piece of code
you **want** to **verify** **made it** into the build."

emmasofiah2o replied:
I think just adding a "when" might fix it. Like so you don’t know where to look for a specific piece of code
when you want to verify that it made it into the build."

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+---
+title: "Embroider Initiative: Progress Update"
+authorHandle: academierenards
```
Judging by the concluding lines of the second-last paragraph, might we want to include more than one author in this post? Or was it just you, @BlueCutOfficial?

BlueCutOfficial replied:
Yes, we could include Chris to make it less weird, his version of the second-last paragraph is more enthusiastic than what I wrote initially XD 
If it's technically possible to have several authors, let's do that 👍

BlueCutOfficial replied:
So with the current code, it's not possible to co-author a blog post, it's only possible to have several mentors for a workshop, but it doesn't rely on the same page template (I don't think it would be hard to implement, so we could suggest that change for a later time).

Do you think I should write the paragraph as a quote? 👇  

<img width="512" alt="Capture d’écran 2024-05-23 à 18 53 27" src="https://github.com/mainmatter/mainmatter.com/assets/22059380/937ea860-130d-4838-84c8-424316bcef87">

emmasofiah2o replied:
I think a quote would look better, yes!

KevinBongart commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+dev mode and build (it essentially just uses Rollup during the build) we need to
+verify that some of the virtualisation features we are implementing work both
+when Ember developers are developing their app and when they are deploying to
+production. Using this new audit system also allows us to more easly investigate
+the contents of the build, e.g. you don’t know how Rollup is going to
```
```suggestion
production. Using this new audit system also allows us to more easily investigate
the content of the build, e.g. you don’t know how Rollup is going to
```

KevinBongart commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+over the past three months and we wouldn't be so close to completing the
+inversion of control work at the time of writing without her!
+
+## Last words
```
```suggestion
## Next steps
```

Maybe less ominous? :D

BlueCutOfficial replied:
But I've already used that expression in "Progress and next steps", my writer's brain doesn't like repetition 🤔 Let me think about it, I'll find something :)

KevinBongart commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,160 @@
+
+## Last words
+
+We are now at the point of wrapping up the technical challenges. We should soon
+have a Vite demo that is a good approximation of the new way to write Ember apps
+and matches the future blueprint. The next phase will be all about communication
+and community work: we want this new fast Embroider + Vite experience to be the
+default experience for all Ember users.
```
This section could use one more link to https://mainmatter.com/embroider-initiative/

BlueCutOfficial commented:
Task list:
- @KevinBongart will get in touch with Marco to clarify the unannounced sponsor issue for the header visual, we are not confident to publish otherwise.
- @emmasofiah2o will check the blog post presentation (I added the images).
- @KevinBongart & @emmasofiah2o will check **the very last paragraph** that I added to answer the suggestions.

I am away on Monday, so if any action has to be done on Monday, please feel free to modify this PR as you like and merge it, you don't need me ;)

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,172 @@
+[Embroider Initiative](https://mainmatter.com/embroider-initiative/), don't miss
+Chris Manson's talk at
+[EmberConf](https://www.emberconf.com/schedule/launching-ember-into-the-future-)
+on May 31th.
```
```suggestion
on May 31st.
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,172 @@
+Using this new audit system also allows us to more easily investigate the
+content of the build, e.g. you want to verify a specific piece of code has been
+included into the build, but you don’t know how Rollup is going to organize the
+files, so you don’t know what file to look in to assert the code is present: the
```
```suggestion
files, so you don’t know what file to look into to verify that the code is present: the
```

emmasofiah2o commented on src/posts/2024-05-24-embroider-update.md:
```diff
@@ -0,0 +1,172 @@
+content of the build, e.g. you want to verify a specific piece of code has been
+included into the build, but you don’t know how Rollup is going to organize the
+files, so you don’t know what file to look in to assert the code is present: the
+audit system is designed to figure this out for you, so you can focus your test
```
```suggestion
audit system is designed to figure this out for you, so you can focus on your test's
```
```suggestion
audit system is designed to figure this out for you, so you can focus your test
```

KevinBongart commented on static/assets/images/posts/2024-05-24-embroider-update/header.jpg:
@emmasofiah2o Would you be able to replace this image with the new one with the Auditboard and Emergent Tree logos?

emmasofiah2o commented:
Can I merge @BlueCutOfficial ?
